lang/step.md

```diff
@@ -359,19 +359,30 @@ impl<M: Memory> Machine<M> {
         };
         let mut locals: Map<LocalName, Place<M>> = default();
 
+        // If the callee might return, then `ret` and `next_block` should be `Some`.
+        // And if it never returns, they are required to be `None`.
+        let fn_might_return = func.ret.is_some();
+        if fn_might_return != ret.is_some() || fn_might_return != next_block.is_some() {
+            throw_ub!("Contradicting information on whether function might return");
+        }
+
         // First evaluate the return place and remember it for `Return`. (Left-to-right!)
-        let (caller_ret_place, caller_ret_abi) = ret;
-        let caller_ret_place = self.eval_place(caller_ret_place)?;
-        // Create place for return local, if needed.
-        if let Some((ret_local, callee_ret_abi)) = func.ret {
+        let caller_ret_place = if fn_might_return {
+            let (caller_ret_place, caller_ret_abi) = ret.unwrap();
+            let caller_ret_place = self.eval_place(caller_ret_place)?;
+            // Create place for return local, if needed.
+            let (ret_local, callee_ret_abi) = func.ret.unwrap();
             let callee_ret_layout = func.locals[ret_local].layout::<M>();
             locals.insert(ret_local, self.mem.allocate(callee_ret_layout.size, callee_ret_layout.align)?);
             if caller_ret_abi != callee_ret_abi {
                 throw_ub!("call ABI violation: return ABI does not agree");
             }
+
+            Some(caller_ret_place)
         } else {
             // FIXME: Can we truly accept any caller ABI if the callee does not return anything?
-        }
+            None
+        };
```

Comment:
I know this feels a bit like a `ret.map`, but there is no `Option::try_map` combinator to account for those `?`.
You know a work-around?